### PR TITLE
 [feat] Control Flow Integrity (CFI) Features

### LINKED ISSUE #157: ROM CFI checks for glitch protection

### PR DESCRIPTION
**Pull request by mhatrevi**
This PR contains the following changes:

 1. CFI library with following features: 
        a. CFI counter to detect out-of-flow code execution. 
        b. XOR based Random number generator. 
        c. Procedural macro to add random delays before and after decorated functions and to detect out-of-flow code execution. 
        d. Variable launder function for assisting with duplicate checks.

 3. Enablement of ROM code with CFI features.

### CODE REVIEW

**Review by fdamato — COMMENTED** [review]
This patch seems to have a very big impact on the debuggability of the code, making it in my opinion much more complex. From the meeting, it wasn`t really clear to me how these changes truly help. Would it possible to include some documentation which would clarify that ?

File: cfi/lib/src/cfi_ctr.rs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    pub fn increment() -> CfiInt {
+        if cfg!(all(feature = "cfi", feature = "cfi-counter")) {
+            let int = Self::read();
+            if !int.is_valid() {
+                cfi_panic(CfiPanicInfo::CounterCorrupt);
+            }
+
+            let (new, overflow) = int.val.overflowing_add(1);
+            if overflow {
+                cfi_panic(CfiPanicInfo::CounterOverflow);
```
Comment (JohnTraverAmd) [inline cfi/lib/src/cfi_ctr.rs:100]: Where do we ensure that we never init the counter to a value where the overflow will happen naturally?

File: cfi/derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    let fn_call = if mod_fn {
+        quote!(#fn_name( #(#param_names,)* ))
+    } else {
+        quote!(Self::#fn_name( #(#param_names,)* ))
+    };
+
+    wrapper_fn.block.stmts.clear();
+    wrapper_fn.block.stmts = parse_quote!(
+        let saved_ctr = caliptra_cfi_lib::CfiCounter::increment();
```
Comment (JohnTraverAmd) [inline cfi/derive/src/lib.rs:64]: Can you describe the attack that you are defending with the counter increment and decrement?

File: cfi/lib/src/cfi_ctr.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    /// Assert the counters are equal
+    #[inline(never)]
+    pub fn assert_eq(val1: CfiInt, val2: CfiInt) {
+        if cfg!(all(feature = "cfi", feature = "cfi-counter")) {
+            if !val1.is_valid() {
+                cfi_panic(CfiPanicInfo::CounterCorrupt);
+            }
+            if !val1.is_valid() {
```
Comment (mschilder123) [inline cfi/lib/src/cfi_ctr.rs:142]: val2, I presume?

File: cfi/lib/tests/test_derive.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    reg.set_s3(4);
+    assert_eq!(reg.s3(), 4);
+}
+
+#[test]
+fn test_rand() {
+    let reg = Xoshiro128Reg;
+    let expected_rand_num: [u32; 10] = [
+        11520, 0, 5927040, 70819200, 2031721883, 1637235492, 1287239034, 3734860849, 3729100597,
```
Comment (mschilder123) [inline cfi/lib/tests/test_derive.rs:89]: Where do these KAT values come from?

File: cfi/lib/src/cfi_ctr.rs (hunk 3 above)
Comment (mhatrevi) [reply cfi/lib/src/cfi_ctr.rs]: Fixed.

File: cfi/derive/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+File Name:
+
+    lib.rs
+
+Abstract:
+
+    File contains CFI procedural macro.
+
+References:
+    https://tf-m-user-guide.trustedfirmware.org/design_docs/tfm_physical_attack_mitigation.html
```
Comment (mschilder123) [inline cfi/derive/src/lib.rs:14]: I do not see counter decrements in their suggestions. Not sure a counter increment/decrement wrapper achieves actual CFI monitoring? Entire calls can still be skipped afaict.

File: rom/dev/src/main.rs [hunk 6]
```diff
@@ -53,6 +54,13 @@ Running Caliptra ROM ...
 pub extern "C" fn rom_entry() -> ! {
     cprintln!("{}", BANNER);
 
+    if !cfg!(feature = "no-cfi") {
```
Comment (RonaldAi) [inline rom/dev/src/main.rs:57]: in other places this feature is called "no_cfi", here it is "no-cfi"

Comment (mhatrevi) [reply rom/dev/src/main.rs:57]: Typo; fixed.

File: cfi/lib/tests/test_derive.rs (hunk 4 above)
Comment (mhatrevi) [reply cfi/lib/tests/test_derive.rs:92]: I used the same reference implementation used in the code: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs

File: cfi/derive/src/lib.rs (hunk 5 above)
Comment (mhatrevi) [reply cfi/derive/src/lib.rs:14]: Based on the discussion we had today, we'll embed either the inc/dec operation in the CFI decorated function and corresponding opposite operation in the wrapper function. This will come as a separate PR.

File: cfi/derive/src/lib.rs (hunk 2 above)
Comment (mhatrevi) [reply cfi/derive/src/lib.rs:64]: The attack we are trying to prevent is of a malicious actor causing a glitch and skipping a CFI decorated function.
I discussed this with Marius today and we agreed that the way this is being done is not helpful from a glitch protection perspective; a more useful approach would be to embed either the inc/dec operation in the CFI decorated function and the corresponding opposite operation in the wrapper function. This will result in a counter mismatch should a CFI decorated function be skipped. This change will come as a separate PR.

File: cfi/lib/src/cfi_ctr.rs (hunk 1 above)
Comment (mhatrevi) [reply cfi/lib/src/cfi_ctr.rs:100]: It's done at the start of rom_entry function in rom/dev/src/main.rs. The counter value that we start with is 0xA5A5A5A5 (0 XORed 0xA5A5A5A5).

File: cfi/lib/src/xoshiro.rs [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+static mut S3: u32 = 0u32;
+
+pub struct Xoshiro128Reg;
+
+impl Xoshiro128Reg {
+    pub fn s0(&self) -> u32 {
+        unsafe {
+            #[cfg(feature = "cfi-test")]
+            {
+                core::ptr::read_volatile(&S0 as *const u32)
```
Comment (korran) [inline cfi/lib/src/xoshiro.rs:38]: While volatile? These values aren't going to change behind the back of the compiler, are they?

File: rom/dev/src/main.rs [hunk 8]
```diff
@@ -54,6 +55,13 @@ Running Caliptra ROM ...
...
+        cprintln!("[state] CFI Enabled");
+        CfiCounter::reset();
```
Comment (korran) [inline rom/dev/src/main.rs:60]: CfiCounter::reset() inidrectly calls Trng::assume_init(), which is a bug (you can't seed from a trng that hasn't been in initialized yet).

File: cfi/lib/src/cfi_ctr.rs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+impl Default for CfiInt {
+    /// Returns the "default value" for a type.
+    fn default() -> Self {
+        Self::encode(0)
+    }
+}
+
+/// CFI counter
+pub enum CfiCounter {}
```
Comment (korran) [inline cfi/lib/src/cfi_ctr.rs:71]: nitpick: What's the point of making this a type? Shouldn't these just be top level functions in the module?

Comment (korran) [reply cfi/lib/src/cfi_ctr.rs:71]: Example use:

```
// nitpick: cfi_counter would be clearer than cfi_ctr.
use cfi::cfi_ctr;

cfi_ctr::reset();
cfi_ctr::corrupt();
```

File: rom/dev/src/main.rs (hunk 8 above)
Comment (korran) [reply rom/dev/src/main.rs]: Fixed.

File: cfi/lib/src/xoshiro.rs (hunk 7 above)
Comment (korran) [reply cfi/lib/src/xoshiro.rs]: Fixed.

File: cfi/lib/src/cfi_ctr.rs [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+                cfi_panic(CfiPanicInfo::CounterMismatch);
+            }
+        }
+    }
+
+    #[inline(never)]
+    pub fn delay() {
+        #[cfg(all(target_arch = "riscv32", feature = "cfi", feature = "cfi-counter"))]
+        unsafe {
+            let cycles = prng().next() % 256;
```
Comment (korran) [inline cfi/lib/src/cfi_ctr.rs:174]: Nitpick: Can you move these first two lines outside the unsafe block? It makes it easier for reviewers to be confident that prng().next() isn't an unsafe method with additional safety requirements.

File: rom/dev/src/flow/cold_reset/fw_processor.rs [hunk 11]
```diff
@@ -170,10 +173,14 @@ impl FirmwareProcessor {
 
         txn.copy_request(slice)?;
 
-        if let Some(result) = ImageManifest::read_from(slice.as_bytes()) {
+        let opt = ImageManifest::read_from(slice.as_bytes());
+        let result = opt.is_some();
+        if cfi_launder(result) {
```
Comment (korran) [inline rom/dev/src/flow/cold_reset/fw_processor.rs:178]: I'm skeptical this is accomplishing much; we really want the code-gen to be executing the opt.is_some() line multiple times.

File: rom/dev/src/flow/update_reset.rs [hunk 12]
```diff
@@ -80,7 +82,11 @@ impl UpdateResetFlow {
         // Populate data vault
         Self::populate_data_vault(venv.data_vault, info);
 
-        Self::load_image(&manifest, recv_txn)?;
+        Self::load_image(&manifest, &mut recv_txn)?;
```
Comment (korran) [inline rom/dev/src/flow/update_reset.rs:85]: What's the purpose of this change?

Comment (korran) [reply rom/dev/src/flow/update_reset.rs:85]: (cfi_impl_fn macro doesn't support mut args).

File: rom/dev/src/pcr.rs [hunk 13]
```diff
@@ -52,111 +56,113 @@ impl PcrExtender<'_> {
...
-) -> CaliptraResult<()> {
-    if pcr_entry_id == PcrLogEntryId::Invalid {
-        return Err(CaliptraError::ROM_GLOBAL_PCR_LOG_INVALID_ENTRY_ID);
+    /// Extend PCR0 and PCR1
+    ///
+    /// # Arguments
+    ///
+    /// * `env` - ROM Environment
+    #[cfg_attr(not(feature = "no-cfi"), cfi_impl_fn)]
+    pub(crate) fn extend_pcrs(
```
[the remark this answers is not in the document]
Comment (korran) [reply rom/dev/src/pcr.rs]: (reverted)

**Review by korran — APPROVED** [review]

File: cfi/lib/src/cfi_ctr.rs (hunk 10 above)
Comment (mhatrevi) [reply cfi/lib/src/cfi_ctr.rs:174]: Will do it as part of the asymmetrical increment and decrement calls in the proc macro.

File: rom/dev/src/flow/cold_reset/fw_processor.rs (hunk 11 above)
Comment (mhatrevi) [reply rom/dev/src/flow/cold_reset/fw_processor.rs:178]: I'll update this as part of the asymm. incr/decr proc. macro changes.

File: cfi/lib/src/cfi_ctr.rs (hunk 9 above)
Comment (mhatrevi) [reply cfi/lib/src/cfi_ctr.rs:81]: We'll make this change as part of the asym incr/decr proc. macro changes.
